text/0000-symbol-name-mangling-v2.md

```diff
@@ -104,25 +206,46 @@ mod foo {
     fn bar'1() {}
 }
 ```
-Then we would encode the two functions symbols as `N15mycrate_4a3b56d3foo3barVE` and `N15mycrate_4a3b56d3foo3barVs_E` respectively (note the `s_` suffix in the second case). Details on the shape of this suffix are provided in the reference-level description.
 
-As opposed to C++ and other languages that support function overloading, we don't need to include function parameter types in the symbol name. Rust does not allow two functions of the same name but different arguments.
+then we would encode the two functions symbols as `NvNtC7mycrate3foo3bar`
+and `NvNtC7mycrate3foos_3bar` respectively (note the `s_` prefix in the
+second case). A very similar disambiguation is needed for avoiding
+conflicts between crates of the same name but different versions. The
+same syntactic prefix is thus used for crate-id where we encode the
+crate disambiguator as in `NtNvCs1234_7mycrate3foo3bar`. Details on
+the shape of this prefix are provided in the reference-level description.
 
-The final symbol name for the function would also include the prefix `_R` that is common to all symbol names generated by this scheme:
-
-```
-  _RN15mycrate_4a3b56d3foo3barVE
-
-  <><-------------------------->
-   |             |
-prefix    absolute path
+As opposed to C++ and other languages that support function overloading,
+we don't need to include function parameter types in the symbol name.
+Rust does not allow two functions of the same name but different arguments.
 
+The final symbol name for the function would also include the prefix
+`_R` that is common to all symbol names generated by this scheme:
 
+```
+  _RNvNtCs1234_7mycrate3foo3bar
+  <>^^^^^<----><------><--><-->
+   ||||||   |      |     |   |
+   ||||||   |      |     |   +--- "bar" identifier
+   ||||||   |      |     +------- "foo" identifier
+   ||||||   |      +------------- "mycrate" identifier
+   ||||||   +-------------------- disambiguator for "mycrate"
+   |||||+------------------------ start-tag for "mycrate"
+   ||||+------------------------- namespace tag for "foo"
+   |||+-------------------------- start-tag for "foo"
+   ||+--------------------------- namespace tag for "bar"
+   |+---------------------------- start-tag for "bar"
```

Comment:
I would maybe describe these slightly differently, i.e. treating the nesting as if
```rust
mycrate::foo::bar
```
was *really*
```rust
((mycrate)::foo)::bar
```
which with the extra information becomes:
```rust
N('v', None, N('t', None, C(Some(1234), "mycrate"), "foo"), "bar")
```